### PR TITLE
Add yield `key` in `Hash#transform_values` and `value` in `#transform_keys`

### LINKED ISSUE #13581: Yield `key` in `Hash#transform_values`
`Hash#transform_values` currently only yields the value that is to be transformed. I'm proposing to also yield the key. It can be useful sometimes and there's not much reason to hide it deliberatly.

This is an example where the key would be used for the transformed value:
```cr
people = {
  "alice" => "alice@example.com",
  "bob" => "bob@example.org",
}

record Contact, name : String, email : String

people.transform_values do |email, name|
  Contact.new(name, email)
end # => {"alice" => Contact(@name="alice", @email="alice@example.com"), "bob" => Contact(@name="bob", @email="bob@example.org")}
```

Implementation should be trivial and cheap. For backwards compatibility, the key would be the second block parameter. This should work with existing code because unused parameters can be omitted in block definitions.

The same applies to `#transform_values!` and I figure `#transform_keys` could be adapted similarly to yield the key as well.

### PR DESCRIPTION
**Pull request by baseballlover723**
Fixes #13581

### CODE REVIEW

File: spec/std/hash_spec.cr [hunk 1]
```diff
@@ -855,6 +855,17 @@ describe "Hash" do
     h2.should be_empty
   end
 
+  it "transforms keys with values included" do
+    h1 = {1 => "a", 2 => "b", 3 => "c"}
+
+    expected_values = ["a", "b", "c"]
```
Comment (Sija) [inline spec/std/hash_spec.cr:861]:
```suggestion
    expected_values = h1.values
```

File: spec/std/hash_spec.cr [hunk 2]
```diff
@@ -878,13 +889,35 @@ describe "Hash" do
     h2.should be_empty
   end
 
+  it "transforms values with keys included" do
+    h1 = {"a" => 1, "b" => 2, "c" => 3}
+
+    expected_keys = ["a", "b", "c"]
```
Comment (Sija) [inline spec/std/hash_spec.cr:895]:
```suggestion
    expected_keys = h1.keys
```

File: spec/std/hash_spec.cr [hunk 3]
```diff
@@ -878,13 +889,35 @@ describe "Hash" do
...
     h = {"a" => 1, "b" => 2, "c" => 3}
 
     h.transform_values!(&.+(1))
     h.should eq({"a" => 2, "b" => 3, "c" => 4})
   end
 
+  it "transform values in place with keys included" do
+    h = {"a" => 1, "b" => 2, "c" => 3}
+
+    expected_keys = ["a", "b", "c"]
```
Comment (Sija) [inline spec/std/hash_spec.cr:913]:
```suggestion
    expected_keys = h1.keys
```

File: spec/std/hash_spec.cr [hunk 4]
```diff
@@ -855,6 +855,17 @@ describe "Hash" do
...
 
+  it "transforms keys with values included" do
+    h1 = {1 => "a", 2 => "b", 3 => "c"}
+
+    expected_values = h1.values
+    h2 = h1.transform_keys do |x, v|
+      v.should eq(expected_values.shift)
+      x + 1
+    end
+    h2.should eq({2 => "a", 3 => "b", 4 => "c"})
```
Comment (straight-shoota) [inline spec/std/hash_spec.cr:866]: Food for thought: Maybe a simpler implementation could combine key and value?
```suggestion
    h1.transform_keys{ |k, v| "#{k}#{v}" }.should eq({"2a" => "a", "3b" => "b", "4c" => "c"})
```
The same can be applied to all other tests.

**Review by beta-ziliani — CHANGES_REQUESTED** [review]
Thanks for the contribution @baseballlover723! Let me add to what was said that the types of the blocks are not respecting the change, and it would be great if the docs can reflect the change too with a little comment.

Let us know if you want us to take over, or if you prefer to take it to completion.

**Comment by baseballlover723** [comment]
@beta-ziliani I'd prefer to take it to completion if possible. To clarify, is what you're asking me to do to update the function defs from `def transform_values(& : V -> V2) : Hash(K, V2) forall V2` -> `def transform_values(& : V, K -> V2) : Hash(K, V2) forall V2`? I was surprised that it worked even without updating the type blocks.

I'll also update the docs as well, though I don't really want to change the example code structure, so the added value will just be unused in the example block.

**Comment by beta-ziliani** [comment]
yes, exactly, that's what I mean. About the documentation, I just meant to add a line stating that the value/key are also yielded to the block

**Comment by baseballlover723** [comment]
@beta-ziliani updated the docs, let me know if the wording is ok with you.

File: src/hash.cr [hunk 5]
```diff
@@ -1611,42 +1611,45 @@ class Hash(K, V)
 
   # Returns a new hash with all keys converted using the block operation.
   # The block can change a type of keys.
+  # The block optionally yields the value as well
```
Comment (straight-shoota) [inline src/hash.cr:1614]: It isn't mentioned previously that the block yields the key, so I think both should be mentioned:
```suggestion
  # The block yields key and value.
```

**Review by beta-ziliani — APPROVED** [review]
Awesome, thanks @baseballlover723 !

File: src/hash.cr [hunk 6]
```diff
@@ -1611,42 +1611,45 @@ class Hash(K, V)
 
   # Returns a new hash with all keys converted using the block operation.
   # The block can change a type of keys.
+  # The block yields the key and value.
   #
   # ```
   # hash = {:a => 1, :b => 2, :c => 3}
-  # hash.transform_keys { |key| key.to_s } # => {"a" => 1, "b" => 2, "c" => 3}
+  # hash.transform_keys { |key, _value| key.to_s } # => {"a" => 1, "b" => 2, "c" => 3}
```
Comment (straight-shoota) [inline src/hash.cr:1618]: This addition isn't really necessary. Unused block parameters can simply be omitted, so I don't think it's useful to write it it with an underscore prefix.
```suggestion
  # hash.transform_keys { |key| key.to_s } # => {"a" => 1, "b" => 2, "c" => 3}
```

Ditto below.

**Review by straight-shoota — APPROVED** [review]

**Pull request merged**